File: Desktop/ClaudeCode/Chess_Games/Endgame_Trainer/old_version/check_bronze.mjs
// check_bronze.mjs — verifies each bronze FEN has a forced mate in ≤12 moves
// Usage: node check_bronze.mjs

import { spawn } from 'child_process';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));

// ── Load positions ───────────────────────────────────────────────────────────
import { createRequire } from 'module';
import vm from 'vm';
const src = readFileSync(join(__dirname, 'positions.js'), 'utf8');
const ctx = {};
vm.createContext(ctx);
vm.runInContext(src, ctx);
const POSITIONS = ctx.POSITIONS;

const TYPE_NAMES = {
  1:  'K+Q vs K',
  2:  'K+R vs K',
  3:  'K+BB vs K',
  4:  'K+BN vs K',
  5:  'K+NN vs K',
  6:  'K+P vs K',
  7:  'Verbygeraakte Pion Wedren',
  8:  'Opposisie & Koningaktiwiteit',
  9:  'Zugzwang',
  10: 'Driehoeksbeweging',
  11: 'Piondeurbraak',
  12: 'Buitenste Verbygeraakte Pion',
  13: 'Lucena-posisie',
  14: 'Philidor-posisie',
  15: 'Toring Agter Verbygeraakte Pion',
  16: 'Aktiewe vs Passiewe Toring',
  17: 'Goeie Loper vs Slegte Loper',
  18: 'Loper teen Ruiter',
  19: 'Verkeerde Kleur Loper',
  20: 'Koningin teen Pion op 7de Ry',
};

// ── Stockfish UCI helper ────────────────────────────────────────────────────
function askStockfish(fen, mateIn) {
  return new Promise((resolve) => {
    const sf = spawn('stockfish');
    let output = '';
    let settled = false;

    const timeout = setTimeout(() => {
      if (!settled) {
        settled = true;
        sf.kill();
        resolve({ found: false, mate: null, timedOut: true });
      }
    }, 10000);

    sf.stdout.on('data', (data) => {
      output += data.toString();
      if (output.includes('\nbestmove ') && !settled) {
        settled = true;
        clearTimeout(timeout);
        sf.kill();
        // Find all "score mate N" occurrences and take the smallest
        const re = /score mate (\d+)/g;
        let m, best = Infinity;
        while ((m = re.exec(output)) !== null) {
          const n = parseInt(m[1]);
          if (n > 0 && n < best) best = n;
        }
        if (best < Infinity) {
          resolve({ found: true, mate: best, timedOut: false });
        } else {
          resolve({ found: false, mate: null, timedOut: false });
        }
      }
    });

    sf.stdin.write('uci\n');
    sf.stdin.write('isready\n');
    sf.stdin.write(`position fen ${fen}\n`);
    sf.stdin.write(`go mate ${mateIn}\n`);
  });
}

// ── Main ────────────────────────────────────────────────────────────────────
async function main() {
  const MATE_LIMIT = 12;
  const results = [];

  console.log(`\nChecking bronze positions for forced mate in ≤${MATE_LIMIT} moves...\n`);
  console.log('─'.repeat(72));

  for (const [typeIdStr, tiers] of Object.entries(POSITIONS)) {
    const typeId = parseInt(typeIdStr);
    const name = TYPE_NAMES[typeId] || `Type ${typeId}`;
    const bronzeList = tiers.bronze || [];

    if (bronzeList.length === 0) {
      console.log(`Type ${typeId}: no bronze positions\n`);
      continue;
    }

    console.log(`Type ${String(typeId).padStart(2)} — ${name}`);

    for (let i = 0; i < bronzeList.length; i++) {
      const { fen } = bronzeList[i];
      process.stdout.write(`  B${i + 1}: ${fen.padEnd(48)} → `);

      const result = await askStockfish(fen, MATE_LIMIT);

      if (result.timedOut) {
        console.log('⏱  TIMEOUT (no forced mate in 10s)');
        results.push({ typeId, name, idx: i + 1, fen, status: 'timeout' });
      } else if (result.found) {
        const ok = result.mate <= MATE_LIMIT;
        console.log(`${ok ? '✅' : '⚠️ '} mate in ${result.mate}${ok ? '' : '  ← OVER LIMIT'}`);
        results.push({ typeId, name, idx: i + 1, fen, status: ok ? 'ok' : 'over', mate: result.mate });
      } else {
        console.log('❌  no forced mate found');
        results.push({ typeId, name, idx: i + 1, fen, status: 'none' });
      }
    }
    console.log();
  }

  // ── Summary ───────────────────────────────────────────────────────────────
  console.log('═'.repeat(72));
  console.log('SUMMARY\n');

  const ok      = results.filter(r => r.status === 'ok');
  const over    = results.filter(r => r.status === 'over');
  const none    = results.filter(r => r.status === 'none');
  const timeout = results.filter(r => r.status === 'timeout');

  console.log(`✅  Confirmed mate ≤12:   ${ok.length}`);
  console.log(`⚠️   Mate found but >12:   ${over.length}`);
  console.log(`❌  No forced mate found:  ${none.length}`);
  console.log(`⏱   Timed out:             ${timeout.length}`);
  console.log(`    Total checked:          ${results.length}`);

  if (over.length || none.length || timeout.length) {
    console.log('\nPositions to review:');
    [...over, ...none, ...timeout].forEach(r => {
      const tag = r.status === 'over' ? `mate in ${r.mate}` : r.status;
      console.log(`  Type ${String(r.typeId).padStart(2)} B${r.idx} [${tag}]`);
      console.log(`         ${r.fen}`);
    });
  } else {
    console.log('\nAll bronze positions pass! 🎉');
  }
}

main().catch(console.error);

File: Desktop/ClaudeCode/Chess_Games/Endgame_Trainer/argief/old_version/check_bronze.py
#!/usr/bin/env python3
"""check_bronze.py — verifies each bronze FEN has a forced mate in ≤12 moves.
Usage: python3 check_bronze.py
Requires: stockfish in PATH, python-chess (pip3 install chess)
"""

import re
import sys
import chess
import chess.engine
from pathlib import Path

POSITIONS_JS = Path(__file__).parent / "positions.js"
MATE_LIMIT   = 12   # full moves (white + black = 1 full move)
SEARCH_TIME  = 15   # seconds per position
WIN_CP       = 200  # centipawns threshold for "clearly winning" (Types 6-20)

# Types 1-5: pure material — must have forced mate ≤ MATE_LIMIT
# Types 6-20: thematic — check for strong winning advantage, mate optional
MATERIAL_TYPES = {1, 2, 3, 4, 5}

TYPE_NAMES = {
    1:  "K+Q vs K",
    2:  "K+R vs K",
    3:  "K+BB vs K",
    4:  "K+BN vs K",
    5:  "K+NN vs K",
    6:  "K+P vs K",
    7:  "Verbygeraakte Pion Wedren",
    8:  "Opposisie & Koningaktiwiteit",
    9:  "Zugzwang",
    10: "Driehoeksbeweging",
    11: "Piondeurbraak",
    12: "Buitenste Verbygeraakte Pion",
    13: "Lucena-posisie",
    14: "Philidor-posisie",
    15: "Toring Agter Verbygeraakte Pion",
    16: "Aktiewe vs Passiewe Toring",
    17: "Goeie Loper vs Slegte Loper",
    18: "Loper teen Ruiter",
    19: "Verkeerde Kleur Loper",
    20: "Koningin teen Pion op 7de Ry",
}


def parse_positions(path):
    src = path.read_text()
    positions = {}
    type_block_re = re.compile(r'^\s{2}(\d+):\s*\{(.*?)\n  \}', re.MULTILINE | re.DOTALL)
    for m in type_block_re.finditer(src):
        type_id = int(m.group(1))
        block = m.group(2)
        positions[type_id] = {}
        tier_re = re.compile(r'(bronze|silver|gold)\s*:\s*\[(.*?)\]', re.DOTALL)
        for tm in tier_re.finditer(block):
            tier = tm.group(1)
            fen_re = re.compile(r'''fen\s*:\s*['"]([^'"]+)['"]''')
            positions[type_id][tier] = fen_re.findall(tm.group(2))
    return positions


def check_fen_material(engine, fen):
    """Types 1-5: check for forced mate within MATE_LIMIT full moves.
    Returns (status, value) where status is 'ok', 'over', or 'none'.
    """
    board = chess.Board(fen)
    try:
        # Search broadly (mate=MATE_LIMIT*2 to handle ply vs move ambiguity)
        info = engine.analyse(
            board,
            chess.engine.Limit(mate=MATE_LIMIT * 2, time=SEARCH_TIME),
        )
        score = info.get("score")
        if score is None:
            return "none", None
        w = score.white()
        if w.is_mate():
            m = w.mate()
            if m is not None and m > 0:
                return ("ok" if m <= MATE_LIMIT else "over"), m
        return "none", None
    except Exception:
        return "none", None


def check_fen_thematic(engine, fen):
    """Types 6-20: check for strong winning advantage.
    Returns (status, value) where status is 'win', 'draw', or 'loss'.
    """
    board = chess.Board(fen)
    try:
        info = engine.analyse(
            board,
            chess.engine.Limit(depth=22, time=SEARCH_TIME),
        )
        score = info.get("score")
        if score is None:
            return "unknown", None
        w = score.white()
        if w.is_mate():
            m = w.mate()
            if m is not None and m > 0:
                return "win", f"mate in {m}"
            else:
                return "loss", f"mated in {abs(m) if m else '?'}"
        cp = w.score()
        if cp is None:
            return "unknown", None
        if cp >= WIN_CP:
            return "win", f"+{cp}cp"
        elif cp <= -WIN_CP:
            return "loss", f"{cp}cp"
        else:
            return "draw", f"{cp:+}cp"
    except Exception:
        return "unknown", None


def main():
    positions = parse_positions(POSITIONS_JS)
    if not positions:
        print("ERROR: Could not parse positions.js", file=sys.stderr)
        sys.exit(1)

    engine = chess.engine.SimpleEngine.popen_uci("stockfish")

    results = []
    print(f"\nChecking bronze positions for forced mate in ≤{MATE_LIMIT} moves...\n")
    print("─" * 72)

    for type_id in sorted(positions.keys()):
        name = TYPE_NAMES.get(type_id, f"Type {type_id}")
        bronze = positions[type_id].get("bronze", [])
        if not bronze:
            print(f"Type {type_id:2}: no bronze positions\n")
            continue

        print(f"Type {type_id:2} — {name}")
        for i, fen in enumerate(bronze):
            print(f"  B{i+1}: {fen:<48} → ", end="", flush=True)
            if type_id in MATERIAL_TYPES:
                status, val = check_fen_material(engine, fen)
                if status == "ok":
                    print(f"✅ mate in {val}")
                elif status == "over":
                    print(f"⚠️  mate in {val}  ← OVER LIMIT")
                else:
                    print("❌  no forced mate found")
            else:
                status, val = check_fen_thematic(engine, fen)
                if status == "win":
                    print(f"✅ winning ({val})")
                elif status == "draw":
                    print(f"⚠️  not clearly winning ({val})")
                elif status == "loss":
                    print(f"❌  losing ({val})")
                else:
                    print("❓  unknown")
            results.append(dict(type_id=type_id, name=name, idx=i+1, fen=fen,
                                status=status, val=val))
        print()

    engine.quit()

    # ── Summary ──────────────────────────────────────────────────────────────
    print("═" * 72)
    print("SUMMARY\n")
    ok_l      = [r for r in results if r["status"] in ("ok", "win")]
    warn_l    = [r for r in results if r["status"] in ("over", "draw")]
    bad_l     = [r for r in results if r["status"] in ("none", "loss")]
    unknown_l = [r for r in results if r["status"] == "unknown"]

    print(f"✅  Pass (mate/winning):    {len(ok_l)}")
    print(f"⚠️   Warning (marginal):    {len(warn_l)}")
    print(f"❌  Fail (no mate/losing):  {len(bad_l)}")
    print(f"❓  Unknown:                {len(unknown_l)}")
    print(f"    Total checked:          {len(results)}")

    problems = warn_l + bad_l + unknown_l
    if problems:
        print("\nPositions to review:")
        for r in problems:
            print(f"  Type {r['type_id']:2} B{r['idx']} [{r['status']}: {r['val']}]")
            print(f"         {r['fen']}")
    else:
        print("\nAll bronze positions pass! 🎉")


if __name__ == "__main__":
    main()
